### PR TITLE
shipit_pulse_listener: Retry retrieving coverage build multiple times before failing and fail without throwing exceptions

### LINKED ISSUE #848: shipit_pulse_listener: Be more robust to failed requests in the code coverage listener
See https://github.com/mozilla-releng/services/issues/827#issuecomment-366204915.

### PR DESCRIPTION
**Pull request by marco-c**
Fixes #848.

### CODE REVIEW

File: lib/cli_common/cli_common/utils.py [hunk 1]
```diff
@@ -19,23 +19,22 @@ def wait_until(operation, timeout=30, interval=30):
 
 
 def retry(operation, retries=5, wait_between_retries=30):
```
Comment (marco-c) [inline lib/cli_common/cli_common/utils.py]: The behavior of this function has been changed to throw an exception when all the retries fail.

File: src/shipit_pulse_listener/shipit_pulse_listener/listener.py [hunk 2]
```diff
@@ -255,23 +256,29 @@ def maybe_trigger(tasks):
...
-            'limit': 200
-        })
-        reply = r.json()
-        task = maybe_trigger(reply['tasks'])
-
-        while task is None and 'continuationToken' in reply:
+        def retrieve_coverage_task():
             r = requests.get(list_url, params={
-                'limit': 200,
-                'continuationToken': reply['continuationToken']
```
Comment (La0) [inline src/shipit_pulse_listener/shipit_pulse_listener/listener.py]: Why the continuationToken has disappeared ?

Comment (La0) [reply src/shipit_pulse_listener/shipit_pulse_listener/listener.py]: Nevermind, that's only for the initial request i guess.

File: src/shipit_pulse_listener/shipit_pulse_listener/listener.py [hunk 3]
```diff
@@ -255,23 +256,27 @@ def maybe_trigger(tasks):
...
+        def retrieve_coverage_task():
+            r = requests.get('https://queue.taskcluster.net/v1/task-group/{}/list'.format(group_id), params={
+                'limit': 200
             })
             reply = r.json()
             task = maybe_trigger(reply['tasks'])
 
-        return task
+            while task is None and 'continuationToken' in reply:
+                r = requests.get(list_url, params={
```
Comment (La0) [inline src/shipit_pulse_listener/shipit_pulse_listener/listener.py]: Where does `list_url` comes from ?

File: src/shipit_pulse_listener/shipit_pulse_listener/listener.py [hunk 4]
```diff
@@ -255,23 +256,27 @@ def maybe_trigger(tasks):
...
+                    'limit': 200,
+                    'continuationToken': reply['continuationToken']
+                })
+                reply = r.json()
```
Comment (La0) [inline src/shipit_pulse_listener/shipit_pulse_listener/listener.py]: Add `r.raise_for_status()` here ..?

File: src/shipit_pulse_listener/shipit_pulse_listener/listener.py [hunk 5]
```diff
@@ -255,23 +256,27 @@ def maybe_trigger(tasks):
...
-        reply = r.json()
-        task = maybe_trigger(reply['tasks'])
-
-        while task is None and 'continuationToken' in reply:
-            r = requests.get(list_url, params={
-                'limit': 200,
-                'continuationToken': reply['continuationToken']
+        def retrieve_coverage_task():
+            r = requests.get('https://queue.taskcluster.net/v1/task-group/{}/list'.format(group_id), params={
+                'limit': 200
```
Comment (La0) [inline src/shipit_pulse_listener/shipit_pulse_listener/listener.py]: Add `r.raise_for_status()` here ..?

File: src/shipit_pulse_listener/shipit_pulse_listener/listener.py (hunk 3 above)
Comment (marco-c) [reply src/shipit_pulse_listener/shipit_pulse_listener/listener.py]: It's defined a few lines earlier: `list_url = 'https://queue.taskcluster.net/v1/task-group/{}/list'.format(group_id)`

File: src/shipit_pulse_listener/shipit_pulse_listener/listener.py (hunk 5 above)
Comment (marco-c) [reply src/shipit_pulse_listener/shipit_pulse_listener/listener.py]: Done with 8f380e2.

File: src/shipit_pulse_listener/shipit_pulse_listener/listener.py (hunk 4 above)
Comment (marco-c) [reply src/shipit_pulse_listener/shipit_pulse_listener/listener.py]: Done with 8f380e2.

File: src/shipit_pulse_listener/shipit_pulse_listener/listener.py (hunk 3 above)
Comment (La0) [reply src/shipit_pulse_listener/shipit_pulse_listener/listener.py]: Okay, i missed it by reading an outdated commit (8a6f8ef7a97249544cc40e300d34567c1a4a847e)

**Pull request merged**